Components: Do not trigger onChange for disabled FormToggle

aduth opened the pull request:
Observed at https://github.com/Automattic/wp-calypso/pull/7047#issuecomment-235691239

This pull request seeks to resolve an issue where `onChange` is still called when clicking a disabled `<FormToggle />` component. This results in dirty form warnings and analytics tracking in the [site writing settings custom type toggles](https://wpcalypso.wordpress.com/settings/writing) when clicking a disabled toggle.

__Implementation notes:__

There are several issues with `<FormToggle />` as a component and is due for some serious refactoring. I attempted to restrain myself with the changes, but see c57af874b3ee5c0fa2a40032c19c7548d784c8ed as an example of changes to be made (that branch isn't working 100% yet).

In this branch, 9e8abc4cbff5f85d51b175da1004d269fcf66550 contains in-place refactoring while 3d757ff4c950b3de6f461f7f9d67592267b886c0 contains the fix itself.

__Testing instructions:__

1. (Prerequisite) Choose a theme for your site which supports Portfolios and/or Testimonials by default (e.g. Sketch)
2. Navigate to [site writing settings](http://calypso.localhost:3000/settings/writing)
3. Select site configured in Step 1
4. Enable analytics debugging by entering `localStorage.debug = 'calypso:analytics';` in your browser developer tools console
5. Refresh the page
6. Click on one of the form toggles
7. Note that...
 - If the toggle is enabled, an analytics event is logged
 - If the toggle is disabled, no analytics event is logged
8. Refresh the page
9. Note that...
 - If the toggle clicked in step 6 is enabled, a dirty form warning is shown
 - If the toggle clicked in step 6 is disabled, no dirty form warning is shown

/cc @timmyc 

Test live: https://calypso.live/?branch=fix/form-toggle-disabled-click

timmyc commented on client/components/forms/form-toggle/index.jsx:
```diff
@@ -1,51 +1,72 @@
+			event.preventDefault();
+			this.props.onChange();
 		}
+
```
Perhaps in your larger refactor adding some default props for `onChange` and `onKeyDown` could remove some of these conditional checks.

timmyc commented:
This fixes the issue I encountered in #7047 - no analytics event is fired if the control is disabled and the form is no longer dirty.  LGTM

aduth commented:
Rebased to bring in fancy new [`React.PureComponent`](https://github.com/facebook/react/blob/master/CHANGELOG.md#1530-july-29-2016), and added default props for event handlers (not just to avoid conditions, but also since technically an error could have been thrown if `onKeyDown` prop was passed by not `onChange`).
